통계프로그래밍 assignment2a

# Scientific Programming in Python - Control Flow & File IO

### Problem 1. FizzBuzz

>**(a)** Write a program that prints out the numbers between one and 100 except that:
* for multiples of three it prints the word Fizz instead of that number;
* for multiples of five it prints the word Buzz instead of that number; and
* for multiples of both three and five it prints the word FizzBuzz instead of that number.

>For example, the first 15 lines of output from your program should be identical to the following:

In [1]:
# Your code here
numbers = range(1,100) #숫자 먼저 정의하고

def FizzBuzz(numbers):
    for i in range(len(numbers)): #지정한 숫자만큼 for문 돈다
        number = numbers[i] #number 하나씩 지정
        if number % 3 ==0 and number % 5 == 0: #3의 배수, 5의 배수 둘다 해당
            print('FizzBuzz') 
        elif number % 3 == 0: # 그 외 3의 배수 해당하는 경우
            print('Fizz')
        elif number % 5 == 0: # 그 외 5의 배수 해당하는 경우
            print('Buzz')
        else:
            print(number)
    
FizzBuzz(numbers[:15])

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz


>**Hint:** If you are having trouble, start by writing code that will print all the numbers one to 100, one per line. Then, modify that code so that it prints the word Fizz instead of multiples of three. Next, modify your code to prints the word Buzz instead of multiples of five. Finally, modify it so that it prints the correct output.

>**Hint:** To test if a variable <code>n</code> is a multiple of three, you can use this code:

>><code>(n % 3) == 0</code>

> The <code>%</code> symbol here is called the [modulo operator](https://en.wikipedia.org/wiki/Modulo_operation) and the code <code>n % 3</code> is pronounced "n modulo three". So if <code>n</code> is zero modulo three, it gives 0 remainder on division by three and that means it’s a multiple of 3. You can use the code like this:

>><code>if (n % 3) == 0:
    do_stuff </code>

>Here, <code>do_stuff</code> will be executed if and only if <code>n</code> is a multiple of 3.

### Problem 2. Library with File I/O

In this problem, you are given a list of student names and library fees in a file and your task is to write some Python code to process it. Make sure you download [library_data.txt](http://ucl-cs-grad.github.io/scipython/notebooks/day2/library_data.txt) from the website and put it in the same folder as your iPython notebook. The file looks something like this:

Each line in the file is a single student record. Each student record has the following format:
><code>StudentNumber FirstName LastName AmountOwed</code>

>**(a)** Write a function that uses a for-loop to iterate over each line in the file and store it in a list. The resulting list should look something like this:

In [2]:
# Your code here

def line_list(file):
    result_list = [] # 빈리스트 먼저 정의
    
    with open(file, "r") as infile: # 파일 불러와서
        for line in infile: #한줄씩 불러오도록 for문
            result_list.append(line) # 파일내에 있는 문장 한줄씩 list에 넣기
            
    return result_list #파일 내 문장들 list로 넣은 것 return 받아오기

file = "library_data.txt"
result = line_list(file)
print(result)

['123456 John Doe 1.49\n', '312314 Jane Miller 0.00\n', '531337 Ted Johnson 8.12\n', '273263 Johnny Depp 0.0\n', '102931 Fred Asteir 1.20\n', '391273 Sarah Connor 10.39\n']


In [3]:
print(result[:3])

['123456 John Doe 1.49\n', '312314 Jane Miller 0.00\n', '531337 Ted Johnson 8.12\n']


>**(b)** Write a function to create two lists:
>1. A list that contains only the first column (the student number)
>2. A list that contains only the remaining information (the student name and the amount of money owed to the library)

>**Hint:** You could use the string method <code>split</code> to split a string into a list of smaller strings. Read the help for this function if needed.

In [4]:
# Your code here
def student_info(file): # list에서 학생 번호와 나머지 정보 분리
    student_number  = [] # 빈리스트 정의
    remain_info = [] # 빈리스트 정의
    
    for info in file: # file에 있는 것 한 줄 씩 불러오고
        number = info.split(" ")[0] #한 리스트 분리 후 첫번째에 해당하는 부분만 가져오기
        student_number.append(number) # number에 해당하는 값은 앞서 정의한 student_number에 넣는다
        
        remain = info.split(" ")[1:] #한 리스트 분리 후 첫번째빼고 나머지에 해당하는 부분만 가져오기
        remain_info.append(remain) #remain에 해당하는 값은 remain.info에 하나씩 더해간다
        
    return student_number, remain_info #완성한 리스트 return 
    
data = line_list(file) # (a)함수로 구한 list 이용
numberlist, remainlist = student_info(data)
print(numberlist)
print(remainlist)

['123456', '312314', '531337', '273263', '102931', '391273']
[['John', 'Doe', '1.49\n'], ['Jane', 'Miller', '0.00\n'], ['Ted', 'Johnson', '8.12\n'], ['Johnny', 'Depp', '0.0\n'], ['Fred', 'Asteir', '1.20\n'], ['Sarah', 'Connor', '10.39\n']]


>**(c)** Write a function that splits the second list futher into two lists:
>1. A list that contains the name
>2. A list that contains the amount owed (as a float)

In [5]:
## Your code here
def name_owed(remainlist): # 이름과 소유하고있는 amount뽑아오기 
    name = [] # 빈리스트 정의
    owed = [] # 빈리스트 정의
    
    for remain in remainlist: #받아온 remainlist에서 하나씩 꺼내온다
        nm = " ".join(remain[:2]) # 위에서 구한 나머지 정보에서 앞의 두부분을 join하면 하나의 이름이 된다
        name.append(nm) #위에서 만든 nm을 name리스트에 하나씩 더해간다
        
        ow = float(remain[2]) # 나머지 부분에서 index 2에 해당하는 값
        owed.append(ow) #ow값도 owed리스트에 더해간다
        
    return name, owed #완성한 리스트 return값으로 지정


numberlist, remainlist = student_info(data) # (b)로 구한 list 사용
name, owed = name_owed(remainlist) # 함수 return값 불러옴
print(name)
print(owed)

['John Doe', 'Jane Miller', 'Ted Johnson', 'Johnny Depp', 'Fred Asteir', 'Sarah Connor']
[1.49, 0.0, 8.12, 0.0, 1.2, 10.39]


>**(d)** Write a function that determines the name and student number of the person that owes the largest amount of money.

In [6]:
# Your code here

def det_largest(owed): # 가장 많은 money를 가지고 있는 사람의 이름과 student number
    max_owed = max(owed) #가장 큰 값 
    index = owed.index(max_owed) # 그 값의 index
    return index #가장많은 money 갖고있는 사람의 index return

name, owed = name_owed(remainlist) # (c)로 구한 list 사용
index = det_largest(owed) 
print("The person that owes the largest amount of money")
print('Name: '+ name[index]) # 해당 index를 이용하여 가장 많은 money갖고 있는 사람의 이름  받아온다
print('student number: ' + str(numberlist[index])) #동일한 방식으로 studen number과
print('amound of money: ' + str(owed[index])) #money도 가져온다

The person that owes the largest amount of money
Name: Sarah Connor
student number: 391273
amound of money: 10.39


>**(e)** Write a function that determines the names and student numbers of all people that owe money (i.e. where the amount is not 0.00).

In [7]:
# Your code here

def det_owepeople(owed,name,numberlist): 
    owe_people = [] # 빈리스트 지정
    for i in range(len(owed)): #
        if owed[i] != 0.00: #소유하고 있는 돈이 0이 아닌 경우 모두 가져옴.
            one_person = [] #빈리스트 지정 (한사람에 대한 리스트를 만드므로) 한 사람이 끝났을 때 다시 빈리스트로 초기화
            one_person.append(name[i]) #해당 사람의 이름
            one_person.append(numberlist[i]) #해당 사람의 student number
            one_person.append(owed[i]) #해당 사람의 소유하고 있는 돈
            owe_people.append(one_person) #하나의 리스트 완성하면 위의 전체 리스트에 더해준다
    return owe_people #리스트 return

owe_people = det_owepeople(owed, name, numberlist)
print('Name / Student_number / amount_of_money   ')
for i in range(len(owe_people)): #받아온 리스트를 한줄씩 돌면서 받아온다
    print(owe_people[i][0] + ' / '+ str(owe_people[i][1])  + ' / '+ str(owe_people[i][2]))    

Name / Student_number / amount_of_money   
John Doe / 123456 / 1.49
Ted Johnson / 531337 / 8.12
Fred Asteir / 102931 / 1.2
Sarah Connor / 391273 / 10.39


>**(f)** We will now write our results to a new file. Write some code that:
> 1. lets the user enter a file name;
> 2. creates this file; and
> 3. writes to the file the list of people that owe money (e.g. one student per line, with how much they owe as well).

In [9]:
# Your code here

file_name = input('Enter a file name: ') #파일이름을 지정할 수 있게 input함수 사용
owe_people = det_owepeople(owed, name, numberlist) #위에서 정의한 함수 이용해서 돈 소유 사람의 list받아온다

def write_newfile(file_name,owe_people): #위에서 구한 내용들 새로운 file로 저장
    with open(file_name, 'w') as f: #파일에 내용을 쓰기위해 파일을 열고
        for people in owe_people: #리스트에서 한줄씩, 즉 한사람씩 줄을 쓴다
            line = ''
            
            for info in people: #한사람이 가지고 있는 정보들 하나씩 
                line = line + str(info) + " "   #하나 쓰고 " "로 공백을 주고 다음거 쓰고 반복함         
            line = line + "\n" #한사람이 끝나면 '\n'으로 다음줄로 넘어간다
            f.write(line) #완성한 line file에 씀

write_newfile(file_name, owe_people)

Enter a file name: info


test.txt 

John Doe 123456 1.49 <br>
Ted Johnson 531337 8.12 <br>
Fred Asteir 102931 1.2 <br>
Sarah Connor 391273 10.39 


### Problem 4. Who’s the winner?

The London elections are around the corner and you have been tasked with writing the code that determines the winner given all the ballots. The winner is determined as follows:

* From the list of candidates (which we will for simplicity number $0,1,2,...$ ), every voter gets to select his first and his second choice. The data you are given can thus be represented as a list of lists:

><code>[ [1st_choice_voter_1, 2nd_choice_voter_1], [1st_choice_voter_2, 2nd_choice_voter_2], ...]</code>

* If any of the candidates gets more than 50% of the first votes, he is declared the winner.


* If no candidate achieves this absolute majority, the two candidates with the most first votes enter into a second round, all others are eliminated.  


* In the second round, all ballots with first votes for candidates that did not enter the second round are re-examined, and any second choice votes for the top two candidates are added to their scores.  


* The candidate with the highest number of combined first and second choice votes is the winner.  

>**(a)** Download the file [votes.pickle](ucl-cs-grad.github.io/scipython/notebooks/day2/votes.pickle) from the course website. Then, use the following code to read the list of votes from the file:

In [10]:
from pickle import * 

votes = load( open("votes.pickle", "rb") ) #저장되어 있는 pickle파일을 불러온다

>**(b)** Fill in the code required to count the first votes and to determine the possible first-round winner. There are 6 candidates in the given data set. You should find that in this data set no candidate wins in the first round.

In [11]:
num_candidates = 6

# create a list to store the number of votes per candidate
first_votes = [0] * num_candidates #미리 list만들어주고, 뒤에서 vote수 count할때 이 list의 해당 index를 후보자의 index라 생각하고 하나씩 더해감

# go through the list of votes and add up the first votes per candidate

### YOUR CODE HERE ###
for (first,second) in votes:
    first_votes[first] += 1 #첫번째 투표만이니까 미리 만들어둔 리스트에 해당 인덱스에 count 해간다

print(first_votes)
# in the first round, if any candidate has > 50% of the votes, he wins

### YOUR CODE HERE ###
for i in range(0, len(first_votes)): #firstvote길이만큼 for문으로 돌면서
    if first_votes[i] > len(votes)/2.0: # 50%이상의 투표수를 가지는지 조건확인
        print('Winner Candidate '+ str(i) + str(first_votes[i]/float(len(votes))) + '%')
        #출력된 값이 없는 것을 보니, 해당하는 위너가 없다.

[10010, 26969, 23130, 4959, 14848, 20084]


>**(c)** Implement the second round process described above. If you do this correctly, you should find that candidate 2 wins with 45619 votes. There are some hints in the skeleton below to get you started

In [12]:
# For the elimination round, we need to find the two candidates with the most first votes.
# The code below does just that: It returns a tuple of the indices of the 
# top two candidates. Don't worry if you don't understand how this works yet --
# we will see a much simpler and more elegant way to achieve the same later in the course.

elimination_candidates  = list( zip ( *sorted( zip( first_votes, range( len(first_votes) ) ) )[-2:] ))[1]

# make a copy so we can add votes
second_votes = first_votes[:] #second vote를 위한 리스트 생성

# add second votes to first votes for remaining candidates

### YOUR CODE HERE ###
for (first, second) in votes: #두번째까지 투표 진행 
    if first not in elimination_candidates:#first에 해당하는 index가 elimination_candidates에 없을때
        if second not in elimination_candidates: #second에도 해당하는 index가 elimination_candidates에 없을 때
            if second != first: #first와 second 다를 경우 -> 이 모든 조건 해당할 때만
                second_votes[second] +=1 #해당 index에 vote count
print(second_votes)

# find and display the winner
# you can either try to adapt the code used to find the top two candidates above, 
# or simply use another for loop

### YOUR CODE HERE ###
winner = sorted( zip(first_votes, range(len(first_votes))))[-1][1]
print(winner) # winner index 번호
print("Winner Candidate index: " + str(winner) )
print("Winner votes : "+ str(second_votes[winner]))

[12021, 26969, 23130, 7234, 16683, 21581]
1
Winner Candidate index: 1
Winner votes : 26969
